src/qibo/gates/gates.py

```diff
@@ -864,6 +868,29 @@ def _dagger(self) -> "Gate":
         theta, lam, phi = tuple(-x for x in self.parameters)  # pylint: disable=E1130
         return self.__class__(self.target_qubits[0], theta, phi, lam)
 
+    def decompose(self) -> List[Gate]:
+        """Decomposition of :math:`U_{3}` up to global phase.
+
+        A global phase difference exists between the definitions of
+        :math:`U3` and this decomposition. More precisely,
+
+        .. math::
+            U_{3}(\\theta, \\phi, \\lambda) = e^{i \\, \\frac{3 \\pi}{2}}
+                \\, \\text{RZ}(\\phi + \\pi) \\, \\sqrt{X} \\, \\text{RZ}(\\theta + \\pi)
+                \\, \\sqrt{X} \\, \\text{RZ}(\\lambda) \\, ,
+
+        where :math:`\\text{RZ}` and :math:`\\sqrt{X}` are, respectively,
+        :class:`qibo.gates.RZ` and :class`qibo.gates.SX`.
+        """
+        q = self.init_args[0]
+        return [
+            RZ(q, self.init_kwargs["lam"]),
+            SX(q),
+            RZ(q, self.init_kwargs["theta"] + math.pi),
+            SX(q),
+            RZ(q, self.init_kwargs["phi"] + math.pi),
+        ]
```

Comment:
Just one small personal doubt: why do you use the two `SX` gates?

I thought that the basic decomposition would have been just Euler angles, i.e. `[RZ(a), RX(b), RZ(c)]`.
Is there any further reason to use 5 gates?

Reply:
That's how it is decomposed for hardware purposes. The two $\sqrt{X}$ are applied as a $\\textup{RX}(\\pi / 2)$ pulse each, and the $\\textup{RZ}$ "applied" via virtual phases.

Reply:
Indeed, that's what the qibolab compiler currently does to U3, but at a pulse level.

Reply:
Ok, thanks for the answer :)

Then I wonder how much this `decompose` feature is overlapping with the compiler, since the choice of the decomposition is also hardware-driven.
However, we can rediscuss the moment the compiler will be moved to Qibo.